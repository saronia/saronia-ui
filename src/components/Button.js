/* @flow */
import styled from 'styled-components';

import { colors, fonts } from '../assets/styles';

type Props = {
  active: boolean,
  inverted: boolean,
  accent: boolean,
};

const Button = styled.button`
  padding: ${props => props.accent ? '0.65rem 1rem 0.35rem 1rem' : '0.5rem 1rem'};
  background-color: ${props => props.inverted ? colors.secondary : colors.primary};
  border-radius: 3px;
  border: none;
  font-size: 1.2rem;
  font-family: ${props => props.accent ? fonts.accent : fonts.system};

  &:hover {
    cursor: pointer;
  }

  &:focus {
    outline: none;
  }
`;

export default Button;
