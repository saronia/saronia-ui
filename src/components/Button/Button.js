/* @flow */
import R from 'ramda';
import React from 'react';
import styled from 'styled-components';
import type { Breed as BreedType, Magnitude as MagnitudeType } from '../../types';
import { Animation, Color, Font, Border, Breed, Magnitude } from '../../assets/constants';

type Props =
  | {|
      accent?: boolean,
      inverted?: boolean,
      onClick?: Function,
      pop?: 'active' | 'focus' | 'hover',
      breed?: BreedType,
      size?: MagnitudeType,
    |}
  | {|
      circular: true,
      radius: MagnitudeType,
      accent?: boolean,
      inverted?: boolean,
      onClick?: Function,
      pop?: 'active' | 'focus' | 'hover',
      pulse?: boolean,
      breed?: BreedType,
      size?: MagnitudeType,
    |};

const getBackgroundColor = R.cond([
  [R.propEq('inverted', true), R.always('transparent')],
  [R.propEq('breed', Breed.SECONDARY), R.always(Color.SECONDARY)],
  [R.T, R.always(colors.primary)],
]);

const getBackgroundHoverColor = R.cond([
  [R.propEq('inverted', true), R.always(colors.whiteDark)],
  [R.propEq('secondary', true), R.always(colors.secondaryDark)],
  [R.T, R.always(colors.primaryDark)],
]);

const getBorder = R.ifElse(
  props => R.isNil(getBorderColor(props)),
  R.always('none'),
  props => R.always(`${getBorderWidth(props)} solid ${getBorderColor(props)}`),
);

const getBorderColor = R.cond([
  [
    R.both(R.propEq('inverted', true), R.propEq('secondary', true)),
    R.always(colors.secondary),
  ],
  [R.propEq('inverted', true), R.always(colors.primary)],
  [R.T, R.always(undefined)],
]);

const getBorderWidth = R.cond([
  [R.propEq('inverted', 'thin'), R.always(borders.width.thin)],
  [R.propEq('inverted', 'medium'), R.always(borders.width.medium)],
  [R.propEq('inverted', 'thick'), R.always(borders.width.thick)],
  [R.T, R.always(borders.width.thin)],
]);

const getColor = R.cond([
  [
    R.both(R.propEq('inverted', true), R.propEq('secondary', true)),
    R.always(colors.secondary),
  ],
  [R.propEq('inverted', true), R.always(colors.primaryDark)],
  [R.propEq('secondary', true), R.always(colors.ivoryDark)],
  [R.T, R.always(colors.black)],
]);

const getFontSize = props => fonts.sizes[R.prop('size', props)];

const getPadding = R.cond([
  [R.propEq('radius', 'mini'), R.always(0)],
  [R.T, R.always('0.2rem 0.5rem')],
]);

const getTransform = R.curry((pseudoState, props) =>
  R.when(R.propEq('pop', pseudoState), R.always('scale(1.13)')),
);

const getWidth = R.cond([
  [R.propEq('radius', 'mini'), R.always('1.5rem')],
  [R.propEq('radius', 'tiny'), R.always('2rem')],
  [R.propEq('radius', 'small'), R.always('2.5rem')],
  [R.propEq('radius', 'medium'), R.always('3rem')],
  [R.propEq('radius', 'large'), R.always('4.5rem')],
  [R.propEq('radius', 'big'), R.always('6rem')],
  [R.propEq('radius', 'huge'), R.always('9rem')],
  [R.propEq('radius', 'massive'), R.always('12rem')],
  [R.T, R.always('inherit')],
]);

const Button = styled.button`
  background-color: ${getBackgroundColor};
  width: ${getWidth};
  height: ${getWidth};
  padding: ${getPadding};
  overflow: hidden;

  ${/* flexbox */ ''}
  display: flex;
  justify-content: ${props => (props.circular ? 'center' : 'flex-start')};

  ${/* font */ ''}
  color: ${getColor};
  font-size: ${getFontSize};
  font-family: ${props => (props.accent ? fonts.accent : fonts.system)};

  ${/* border */ ''}
  border: ${getBorder};
  border-radius: ${props => (props.circular ? '50%' : borders.radius)};

  ${/* transition */ ''}
  transition: all .3s, transform .4s ease-out;

  &:hover {
    cursor: pointer;
    background-color: ${getBackgroundHoverColor};
    transform: ${getTransform('hover')};
  }

  &:active {
    transform: ${getTransform('active')};
  }

  &:focus {
    transform: ${getTransform('focus')};
    outline: none;
  }
`;

const Pulse = styled.div`
  position: relative;
`;

const Overlay = styled(Button)`
  position: absolute;
  top: 0;
  left: 0;
  color: transparent;
  z-index: -1;
  animation-name: ${animations.scaleUpAndFadeOut};
  animation-duration: 1.6s;
  animation-iteration-count: infinite;
  animation-delay: .2s;
`;

export default (props: Props) => {
  return R.cond([
    [
      R.propEq('pulse', true),
      R.always(
        <Pulse>
          <Button {...props} />
          <Overlay {...R.merge(props, { inverted: true })} />
        </Pulse>,
      ),
    ],
    [R.T, R.always(<Button {...props} />)],
  ])(props);
};
