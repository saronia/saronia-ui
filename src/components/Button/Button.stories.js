/* @flow */
import R from 'ramda';
import React from 'react';
import { storiesOf, action, linkTo } from '@kadira/storybook';
import {
  withKnobs,
  text,
  boolean,
  number,
  select,
} from '@kadira/storybook-addon-knobs';

import { Button } from '.';
import { SizeEnum } from '../../assets/constants';

const stories = storiesOf('Button', module);

stories.addDecorator(withKnobs);
stories.addDecorator(story => (
  <div style={{ margin: '50px' }}>
    {story()}
  </div>
));

const popOptions = R.invertObj({
  ACTIVE: 'active',
  FOCUS: 'focus',
  HOVER: 'hover',
  NONE: undefined,
});

stories.add('default', () => {
  const circular = boolean('circular', false);
  const props = {
    secondary: boolean('secondary', false),
    accent: boolean('accent', false),
    inverted: boolean('inverted', false),
    size: select('size', R.invertObj(SizeEnum), 'medium'),
    children: text('children', 'Register'),
    circular,
    radius: circular
      ? select('radius', R.invertObj(SizeEnum), 'large')
      : undefined,
    pop: select('pop', popOptions, undefined),
    pulse: circular ? boolean('pulse', false) : undefined,
  };

  return R.cond([
    [
      R.propEq('circular', true),
      R.always(
        <Button
          accent={props.accent}
          circular
          inverted={props.inverted}
          onClick={action('clicked')}
          pop={props.pop}
          pulse={props.pulse}
          radius={props.radius || 'mini'}
          secondary={props.secondary}
          size={props.size}
        >
          {props.children}
        </Button>,
      ),
    ],
    [
      R.T,
      R.always(
        <Button
          accent={props.accent}
          inverted={props.inverted}
          onClick={action('clicked')}
          pop={props.pop}
          secondary={props.secondary}
          size={props.size}
        >
          {props.children}
        </Button>,
      ),
    ],
  ])(props);
});
