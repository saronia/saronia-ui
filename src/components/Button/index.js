/* @flow */
import R from 'ramda';
import React from 'react';
import styled from 'styled-components';

import { colors, fonts, borders } from '../../assets/styles';

type Props = {
  inverted?: boolean | 'thin' | 'medium' | 'thick',
  accent?: boolean,
  secondary?: boolean,
};

const getBackgroundColor = R.cond([
  [R.prop('inverted'), R.always(colors.transparent)],
  [R.prop('secondary'), R.always(colors.secondary)],
  [R.T, R.always(colors.primary)],
]);

const getBackgroundHoverColor = R.cond([
  [R.prop('inverted'), R.always(colors.whiteDark)],
  [R.prop('secondary'), R.always(colors.secondaryDark)],
  [R.T, R.always(colors.primaryDark)],
]);

const getBorderWidth = R.cond([
  [R.propEq('inverted', 'thin'), R.always(borders.width.thin)],
  [R.propEq('inverted', 'medium'), R.always(borders.width.medium)],
  [R.propEq('inverted', 'thick'), R.always(borders.width.thick)],
  [R.T, R.always(borders.width.thin)],
]);

const getBorderColor = R.cond([
  [
    R.both(R.prop('inverted'), R.prop('secondary')),
    R.always(colors.secondary),
  ],
  [R.prop('inverted'), R.always(colors.primary)],
  [R.T, R.always(undefined)],
]);

const getBorder = R.ifElse(
  props => R.isNil(getBorderColor(props)),
  R.always('none'),
  props => R.always(`${getBorderWidth(props)} solid ${getBorderColor(props)}`),
);

const getColor = R.cond([
  [R.both(R.prop('inverted'), R.prop('secondary')), R.always(colors.secondary)],
  [R.prop('secondary'), R.always(colors.ivoryDark)],
  [R.T, R.always(colors.black)],
]);

const Button = styled.button`
  padding: ${props => props.accent ? '0.65rem 1rem 0.35rem 1rem' : '0.5rem 1rem'};
  background-color: ${getBackgroundColor};
  font-size: 1.2rem;
  font-family: ${props => props.accent ? fonts.accent : fonts.system};
  color: ${getColor};
  transition-duration: 0.25s;

  border: ${getBorder};
  border-radius: ${borders.radius};

  &:hover {
    cursor: pointer;
    background-color: ${getBackgroundHoverColor};
  }

  &:focus {
    outline: none;
  }
`;

export default (props: Props) => <Button {...props} />;
