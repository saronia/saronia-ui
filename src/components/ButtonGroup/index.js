/* @flow */
import React from 'react';
import styled from 'styled-components';
import { borders } from '../../assets/styles';

type Props = {
  vertical?: boolean,
};

const ButtonGroup = styled.div`
  display: inline-flex;
  flex-direction: ${props => props.vertical && 'column' || 'row'};

  button {
    border-radius: 0;

    &:first-child {
      border-top-left-radius: ${borders.radius};
      border-top-right-radius: ${props => props.vertical ? borders.radius : 0};
      border-bottom-left-radius: ${props => props.vertical ? 0 : borders.radius};
    }

    &:last-child {
      border-top-right-radius: ${props => props.vertical ? 0 : borders.radius};
      border-bottom-right-radius: ${borders.radius};
      border-bottom-left-radius: ${props => props.vertical ? borders.radius : 0};
    }
  }
`;

export default (props: Props) => <ButtonGroup {...props} />;
