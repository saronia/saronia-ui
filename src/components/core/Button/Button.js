/* @flow */
import React, { Children } from 'react';
import R from 'ramda';
import type { Theme, Size } from '../../../lib/types';

import RawButton from './RawButton';
import IconButton from './IconButton';

export type DefaultButtonProps = {
  accent?: boolean,
  circular?: false,
  compact?: false,
  inverted?: boolean,
  iconLeft?: string,
  iconRight?: string,
  onClick?: Function,
  pop?: 'active' | 'focus' | 'hover',
  size?: Size,
  theme?: Theme,
  children?: string | Children,
};

export type CircularButtonProps = {
  accent?: boolean,
  circular: true,
  compact?: false,
  inverted?: boolean,
  onClick?: Function,
  pop?: 'active' | 'focus' | 'hover',
  radius: Size,
  size?: Size,
  theme?: Theme,
  children?: string | Children,
};

export type Props = DefaultButtonProps | CircularButtonProps;

export default (props: Props) =>
  R.cond([
    [
      R.prop('iconLeft'),
      R.always(<IconButton {...R.omit(['iconRight'], props)} />),
    ],
    [
      R.prop('iconRight'),
      R.always(<IconButton {...R.omit(['iconLeft'], props)} />),
    ],
    [
      R.T,
      R.always(
        <RawButton {...props}>
          {React.Children.map(
            props.children,
            child =>
              React.isValidElement(child)
                ? React.cloneElement(child, {
                  accent: props.accent,
                  theme: props.theme,
                  inverted: !props.inverted,
                  size: props.size,
                })
                : child,
          )}
        </RawButton>,
      ),
    ],
  ])(props);
