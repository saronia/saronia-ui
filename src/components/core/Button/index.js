/* @flow */
export { default as Button } from './Button';
export { default as RawButton } from './RawButton';
export { default as LinkButton } from './LinkButton';
export { default as PulseButton } from './PulseButton';
export { default as Buttons } from './Buttons';
