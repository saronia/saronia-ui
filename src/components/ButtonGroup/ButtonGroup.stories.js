/* @flow */
import React from 'react';
import { storiesOf, action, linkTo } from '@kadira/storybook';
import {
  withKnobs,
  text,
  boolean,
  number,
} from '@kadira/storybook-addon-knobs';

import ButtonGroup from '.';
import Button from '../Button';

const stories = storiesOf('ButtonGroup', module);
stories.addDecorator(withKnobs);

stories
  .add('default', () => (
    <ButtonGroup>
      <Button>Left</Button>
      <Button>Middle</Button>
      <Button>Right</Button>
    </ButtonGroup>
  ))
  .add('veritcal', () => (
    <ButtonGroup vertical>
      <Button>Left</Button>
      <Button>Middle</Button>
      <Button>Right</Button>
    </ButtonGroup>
  ))
