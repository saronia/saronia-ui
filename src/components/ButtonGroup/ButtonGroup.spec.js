/* @flow */
import React from 'react';
import renderer from 'react-test-renderer';
import { matcher, serializer } from 'jest-styled-components';

import ButtonGroup from '.';
import Button from '../Button';

expect.addSnapshotSerializer(serializer);
expect.extend(matcher);

describe('<ButtonGroup />', () => {
  it('renders a div w/o crashing', () => {
    const tree = renderer.create(
      <ButtonGroup>
        <Button>Left</Button>
        <Button>Right</Button>
      </ButtonGroup>,
    ).toJSON();

    expect(tree.type).toEqual('div');
  });
});

describe('<ButtonGroup horizontal />', () => {
  const tree = renderer.create(
    <ButtonGroup>
      <Button>Left</Button>
      <Button>Middle</Button>
      <Button>Right</Button>
    </ButtonGroup>,
  ).toJSON();

  it('first button has no border-radius on right corners', () => {
    expect(tree).toMatchStyledComponentsSnapshot();
  });

  it('last button has no border-radius on left corners');

  it('in-between buttons have no border-radius in any corner');
});

// describe('<ButtonGroup vertical />', () => {
//   it('has no border-radius on right corners of first button');
//
//   it('has no border-radius on left corners of last button');
//
//   it('in-between buttons have no border-radius in any corner');
// });
