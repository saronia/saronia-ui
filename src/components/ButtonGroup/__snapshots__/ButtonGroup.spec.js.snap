// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup horizontal /> matches styled component snapshot 1`] = `
.kkztTS {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-inline-flexbox;
  display: -webkit-inline-flex;
  display: inline-flex;
  flex-direction: row;
  -webkit-box-direction: normal;
  -webkit-box-orient: horizontal;
  -ms-flex-direction: row;
  -webkit-flex-direction: row;
}

.kkztTS button {
  border-radius: 0;
}

.kkztTS button:first-child {
  border-top-left-radius: 3px;
  border-top-right-radius: 0;
  border-bottom-left-radius: 3px;
}

.kkztTS button:last-child {
  border-top-right-radius: 3px;
  border-bottom-right-radius: 3px;
  border-bottom-left-radius: 0;
}

.gVHuyA {
  background-color: #FFB00C;
  width: inherit;
  height: inherit;
  padding: 0.2rem 0.5rem;
  overflow: hidden;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: flex-start;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  color: #161616;
  font-size: ;
  font-family: 'Quicksand', sans-serif;
  border: none;
  border-radius: 3px;
  transition: all .3s, -ms-transform .4s ease-out, -webkit-transform .4s ease-out, transform .4s ease-out;
  -webkit-transition: all .3s, -webkit-transform .4s ease-out, transform .4s ease-out;
}

.gVHuyA:hover {
  cursor: pointer;
  background-color: #FFA104;
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -left: ;
}



.gVHuyA:active {
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -left: ;
}



.gVHuyA:focus {
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -left: ;
  outline: none;
}



.eeMFOQ {
  background-color: #FFB00C;
  width: inherit;
  height: inherit;
  padding: 0.2rem 0.5rem;
  overflow: hidden;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: flex-start;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  color: #161616;
  font-size: ;
  font-family: 'Quicksand', sans-serif;
  border: none;
  border-radius: 3px;
  transition: all .3s, -ms-transform .4s ease-out, -webkit-transform .4s ease-out, transform .4s ease-out;
  -webkit-transition: all .3s, -webkit-transform .4s ease-out, transform .4s ease-out;
}

.eeMFOQ:hover {
  cursor: pointer;
  background-color: #FFA104;
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -middle: ;
}



.eeMFOQ:active {
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -middle: ;
}



.eeMFOQ:focus {
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -middle: ;
  outline: none;
}



.eEIXmf {
  background-color: #FFB00C;
  width: inherit;
  height: inherit;
  padding: 0.2rem 0.5rem;
  overflow: hidden;
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: flex-start;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  color: #161616;
  font-size: ;
  font-family: 'Quicksand', sans-serif;
  border: none;
  border-radius: 3px;
  transition: all .3s, -ms-transform .4s ease-out, -webkit-transform .4s ease-out, transform .4s ease-out;
  -webkit-transition: all .3s, -webkit-transform .4s ease-out, transform .4s ease-out;
}

.eEIXmf:hover {
  cursor: pointer;
  background-color: #FFA104;
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -right: ;
}



.eEIXmf:active {
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -right: ;
}



.eEIXmf:focus {
  transform: children:;
  -ms-transform: children:;
  -webkit-transform: children:;
  -right: ;
  outline: none;
}



<div
  className="kkztTS"
>
  <button
    className="gVHuyA"
  >
    Left
  </button>
  <button
    className="eeMFOQ"
  >
    Middle
  </button>
  <button
    className="eEIXmf"
  >
    Right
  </button>
</div>
`;
